btc-types/src/u256.rs

```diff
@@ -0,0 +1,310 @@
+use std::ops::{Div, Not, Rem, Shl, Shr};
+
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+pub struct U256(u128, u128);
+
+impl U256 {
+    const MAX: U256 = U256(
+        0xffff_ffff_ffff_ffff_ffff_ffff_ffff_ffff,
+        0xffff_ffff_ffff_ffff_ffff_ffff_ffff_ffff,
+    );
+
+    const ZERO: U256 = U256(0, 0);
+
+    const ONE: U256 = U256(0, 1);
+
+    /// Creates `U256` from a big-endian array of `u8`s.
+    #[cfg_attr(all(test, mutate), mutate)]
+    pub fn from_be_bytes(a: &[u8; 32]) -> U256 {
+        let (high, low) = split_in_half(a);
+        let big = u128::from_be_bytes(high);
+        let little = u128::from_be_bytes(low);
+        U256(big, little)
+    }
+
+    /// Creates a `U256` from a little-endian array of `u8`s.
+    #[cfg_attr(all(test, mutate), mutate)]
+    pub fn from_le_bytes(a: &[u8; 32]) -> U256 {
+        let (high, low) = split_in_half(a);
+        let little = u128::from_le_bytes(high);
+        let big = u128::from_le_bytes(low);
+        U256(big, little)
+    }
+
+    /// Converts `U256` to a big-endian array of `u8`s.
+    #[cfg_attr(all(test, mutate), mutate)]
+    pub fn to_be_bytes(self) -> [u8; 32] {
+        let mut out = [0; 32];
+        out[..16].copy_from_slice(&self.0.to_be_bytes());
+        out[16..].copy_from_slice(&self.1.to_be_bytes());
+        out
+    }
+
+    /// Calculates 2^256 / (x + 1) where x is a 256 bit unsigned integer.
+    ///
+    /// 2**256 / (x + 1) == ~x / (x + 1) + 1
+    ///
+    /// (Equation shamelessly stolen from bitcoind)
+    pub fn inverse(&self) -> U256 {
+        // We should never have a target/work of zero so this doesn't matter
+        // that much but we define the inverse of 0 as max.
+        if self.is_zero() {
+            return U256::MAX;
+        }
+        // We define the inverse of 1 as max.
+        if self.is_one() {
+            return U256::MAX;
+        }
+        // We define the inverse of max as 1.
+        if self.is_max() {
+            return U256::ONE;
+        }
+
+        let ret = !*self / self.wrapping_inc();
+        ret.wrapping_inc()
+    }
+
+    #[cfg_attr(all(test, mutate), mutate)]
+    fn is_zero(&self) -> bool {
+        self.0 == 0 && self.1 == 0
+    }
+
+    #[cfg_attr(all(test, mutate), mutate)]
+    fn is_one(&self) -> bool {
+        self.0 == 0 && self.1 == 1
+    }
+
+    #[cfg_attr(all(test, mutate), mutate)]
+    fn is_max(&self) -> bool {
+        self.0 == u128::MAX && self.1 == u128::MAX
+    }
+
+    /// Returns the least number of bits needed to represent the number.
+    #[cfg_attr(all(test, mutate), mutate)]
+    fn bits(&self) -> u32 {
```

Comment:
I would like to avoid custom changes over the original implementation https://github.com/rust-bitcoin/rust-bitcoin/blob/d5149c1f5445b28e73a38b58e9a864b0709c3c78/bitcoin/src/pow.rs#L482